File: static/src/api/evidence.ts
/**
 * Evidence Module API Client
 * 证据整理模块全部端点封装
 */
import * as api from './client'

// ─── 类型定义 ─────────────────────────────────────────────────────────────────

export type CaseType = 'injury' | 'death' | 'neonatal'

export interface MaterialResponse {
  id: string
  original_filename: string | null
  file_type: string
  minio_bucket: string | null
  minio_key: string | null
  file_size: number | null
  auto_category: string | null
  manual_category: string | null
  effective_category: string | null
  category_confidence: number | null
  ocr_status: string
  ocr_text: string | null
  ocr_result: Record<string, unknown>
  page_count: number | null
  selected_pages: number[]
  extracted_data: Record<string, unknown>
  manual_edit: Record<string, unknown>
  catalog_index: number | null
  catalog_title: string | null
  catalog_description: string | null
  proof_purpose: string | null
  fee_detail: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface StepResponse {
  id: number
  step_name: string
  status: string
  progress: number
  duration_ms: number | null
  error_message: string | null
  started_at: string | null
  completed_at: string | null
}

export interface EvidenceCase {
  id: string
  case_name: string
  case_type: string
  is_minor: boolean
  status: string
  complaint_case_id: string | null
  plaintiff_info: Record<string, unknown>
  defendant_info: Record<string, unknown>
  catalog_data: Record<string, unknown>
  catalog_pdf_path: string | null
  analysis_result: Record<string, unknown>
  validation_result: Record<string, unknown>
  missing_items: Record<string, unknown>
  export_bundle_path: string | null
  export_files: Record<string, unknown>
  lawyer_info: { name: string; phone: string }[]
  metadata: Record<string, unknown>
  materials: MaterialResponse[]
  steps: StepResponse[]
  created_at: string
  updated_at: string
}

export interface EvidenceCaseListItem {
  id: string
  case_name: string
  case_type: string
  is_minor: boolean
  status: string
  created_at: string
  updated_at: string
}

export interface EvidenceCaseListResponse {
  items: EvidenceCaseListItem[]
  total: number
}

export interface ProgressResponse {
  case_id: string
  status: string
  current_step: string | null
  total_steps: number
  completed_steps: number
  progress_percent: number
  steps: StepResponse[]
}

export interface CatalogGroup {
  category: string
  category_name: string
  items: MaterialResponse[]
}

export interface CatalogResponse {
  case_id: string
  case_name: string
  case_type: string
  groups: CatalogGroup[]
  fee_summary: Record<string, unknown>
  total_amount: number
}

export interface AnalysisResponse {
  case_id: string
  status: string
  analysis_result: Record<string, unknown>
  validation_result: Record<string, unknown>
  missing_items: Record<string, unknown>
  error_message: string | null
}

export interface ProcessResponse {
  case_id: string
  message: string
  task_id: string | null
}

export interface ExportBundleResponse {
  case_id: string
  message: string
  bundle_path: string | null
}

// ─── API 调用 ─────────────────────────────────────────────────────────────────

/** 创建证据案件 */
export async function createCase(data: {
  case_name: string
  case_type: CaseType
  is_minor?: boolean
}): Promise<EvidenceCase> {
  return api.post<EvidenceCase>('/evidence/cases', data)
}

/** 获取案件列表 */
export async function listCases(page = 1, size = 20): Promise<EvidenceCaseListResponse> {
  return api.get<EvidenceCaseListResponse>('/evidence/cases', { page: String(page), size: String(size) })
}

/** 获取案件详情 */
export async function getCase(caseId: string): Promise<EvidenceCase> {
  return api.get<EvidenceCase>(`/evidence/cases/${caseId}`)
}

/** 更新案件基本信息 */
export async function updateCase(caseId: string, data: {
  case_name?: string
  case_type?: CaseType
  is_minor?: boolean
  lawyer_info?: { name: string; phone: string }[]
  defendant_phone?: string
}): Promise<EvidenceCase> {
  return api.put<EvidenceCase>(`/evidence/cases/${caseId}`, data)
}

/** 删除案件 */
export async function deleteCase(caseId: string): Promise<{ message: string }> {
  return api.del(`/evidence/cases/${caseId}`)
}

/** 上传原始素材 */
export async function uploadMaterials(caseId: string, files: File[]): Promise<MaterialResponse[]> {
  const form = new FormData()
  for (const file of files) {
    form.append('files', file)
  }
  const key = import.meta.env.VITE_API_KEY
  const h: Record<string, string> = {}
  if (key) h['X-API-Key'] = key

  const res = await fetch(`/api/v1/evidence/cases/${caseId}/upload`, {
    method: 'POST',
    headers: h,
    body: form,
  })

  if (!res.ok) {
    const body = await res.json().catch(() => ({}))
    throw new Error(body?.detail || `Upload failed: HTTP ${res.status}`)
  }

  return res.json()
}

/** 开始处理（OCR + 分类 + 目录生成） */
export async function processCase(caseId: string): Promise<ProcessResponse> {
  return api.post<ProcessResponse>(`/evidence/cases/${caseId}/process`)
}

/** 获取处理进度 */
export async function getProgress(caseId: string): Promise<ProgressResponse> {
  return api.get<ProgressResponse>(`/evidence/cases/${caseId}/progress`)
}

/** 获取证据目录 */
export async function getCatalog(caseId: string): Promise<CatalogResponse> {
  return api.get<CatalogResponse>(`/evidence/cases/${caseId}/catalog`)
}

/** 更新证据目录 */
export async function updateCatalog(caseId: string, items: {
  material_id: string
  manual_category?: string
  catalog_title?: string
  catalog_description?: string
  proof_purpose?: string
  sort_order?: number
}[]): Promise<void> {
  await api.put(`/evidence/cases/${caseId}/catalog`, { items })
}

/** 更新单个材料 */
export async function updateMaterial(caseId: string, materialId: string, data: {
  manual_category?: string
  catalog_title?: string
  catalog_description?: string
  proof_purpose?: string
  manual_edit?: Record<string, unknown>
}): Promise<MaterialResponse> {
  return api.put<MaterialResponse>(`/evidence/cases/${caseId}/materials/${materialId}`, data)
}

/** 删除材料 */
export async function deleteMaterial(caseId: string, materialId: string): Promise<void> {
  await api.del(`/evidence/cases/${caseId}/materials/${materialId}`)
}

/** 重试单个素材的 OCR */
export async function retryMaterialOcr(caseId: string, materialId: string): Promise<{ message: string }> {
  return api.post<{ message: string }>(`/evidence/cases/${caseId}/materials/${materialId}/retry-ocr`)
}

/** 导出目录 PDF（表格形式） */
export async function exportCatalogPdf(caseId: string): Promise<void> {
  await api.downloadBlob(`/evidence/cases/${caseId}/catalog/pdf`, '证据目录.pdf')
}

/** 导出证据材料 PDF（图片网格排版） */
export async function exportMaterialsPdf(caseId: string): Promise<void> {
  await api.downloadBlob(`/evidence/cases/${caseId}/materials/pdf`, '证据材料.pdf')
}

/** 开始分析 */
export async function analyzeCase(caseId: string): Promise<ProcessResponse> {
  return api.post<ProcessResponse>(`/evidence/cases/${caseId}/analyze`)
}

/** 获取分析结果 */
export async function getAnalysis(caseId: string): Promise<AnalysisResponse> {
  return api.get<AnalysisResponse>(`/evidence/cases/${caseId}/analysis`)
}

/** 导出立案证据 */
export async function exportFilingEvidence(caseId: string): Promise<void> {
  await api.downloadBlob(`/evidence/cases/${caseId}/export/filing-evidence`, '立案证据.docx')
}

/** 导出民事起诉状 */
export async function exportComplaint(caseId: string): Promise<void> {
  await api.downloadBlob(`/evidence/cases/${caseId}/export/complaint`, '民事起诉状.docx')
}

/** 导出司法鉴定申请书 */
export async function exportAppraisalApp(caseId: string): Promise<void> {
  await api.downloadBlob(`/evidence/cases/${caseId}/export/appraisal-app`, '司法鉴定申请书.docx')
}

/** 导出赔偿费用汇总 */
export async function exportCompensation(caseId: string): Promise<void> {
  await api.downloadBlob(`/evidence/cases/${caseId}/export/compensation`, '赔偿费用清单.xlsx')
}

/** 导出指定类型费用明细 */
export async function exportFeeTypeDetail(caseId: string, feeType: string): Promise<void> {
  await api.downloadBlob(`/evidence/cases/${caseId}/export/compensation/${feeType}`, `${feeType}.xlsx`)
}

/** 一键打包导出（同步下载 ZIP） */
export async function exportBundle(caseId: string): Promise<void> {
  await api.downloadBlob(`/evidence/cases/${caseId}/export/bundle/download`, '立案立档包.zip')
}

// ─── 多页文档三步工作流 ─────────────────────────────────────────────────────

export interface PagePreview {
  page: number
  width: number
  height: number
  thumbnail_b64: string
  note?: string
}

export interface PagePreviewResponse {
  material_id: string
  file_type: string
  total_pages: number
  selected_pages: number[]
  pages: PagePreview[]
}

/** 第一步：预览多页文档全部页面（缩略图） */
export async function previewMaterialPages(
  caseId: string,
  materialId: string
): Promise<PagePreviewResponse> {
  return api.get<PagePreviewResponse>(
    `/evidence/cases/${caseId}/materials/${materialId}/pages/preview`
  )
}

/** 第二步：选择需要处理的目标页（空列表=处理全部） */
export async function selectMaterialPages(
  caseId: string,
  materialId: string,
  selectedPages: number[]
): Promise<{
  material_id: string
  selected_pages: number[]
  message: string
}> {
  return api.post(
    `/evidence/cases/${caseId}/materials/${materialId}/pages/select`,
    { selected_pages: selectedPages }
  )
}

/** 第三步：提取指定页为高清图像 URL（用于查看或下载） */
export function getExtractPageUrl(
  caseId: string,
  materialId: string,
  pageNum: number,
  dpi = 150
): string {
  const base = import.meta.env.VITE_API_BASE || '/api/v1'
  return `${base}/evidence/cases/${caseId}/materials/${materialId}/pages/${pageNum}/extract?dpi=${dpi}`
}

File: api/schemas/evidence.py
"""
证据模块 API Schema
"""
from __future__ import annotations

import re
from datetime import datetime
from decimal import Decimal
from typing import Any, List, Literal, Optional

from pydantic import BaseModel, Field, field_validator


# ─── 创建案件 ────────────────────────────────────────────────────────────────

class CreateEvidenceCaseRequest(BaseModel):
    case_name: str = Field(..., min_length=1, max_length=500, description="案件名称")
    case_type: Literal["injury", "death"] = Field(..., description="案件类型: injury=医疗损害（伤残）, death=医疗损害（死亡）")
    is_minor: bool = Field(default=False, description="是否未成年人")
    # 原被告信息从素材中自动提取，创建时无需手动填写
    plaintiff_info: dict[str, Any] = Field(default_factory=dict, description="原告信息（自动提取）")
    defendant_info: dict[str, Any] = Field(default_factory=dict, description="被告信息（自动提取）")


# ─── 案件响应 ────────────────────────────────────────────────────────────────

class MaterialResponse(BaseModel):
    id: str
    original_filename: str | None = None
    file_type: str = "other"
    minio_bucket: str | None = None
    minio_key: str | None = None
    file_size: int | None = None
    auto_category: str | None = None
    manual_category: str | None = None
    effective_category: str | None = None
    category_confidence: float | None = None
    ocr_status: str = "pending"
    ocr_text: str | None = None
    ocr_result: dict[str, Any] = Field(default_factory=dict)
    page_count: int | None = None
    extracted_data: dict[str, Any] = Field(default_factory=dict)
    manual_edit: dict[str, Any] = Field(default_factory=dict)
    catalog_index: int | None = None
    catalog_title: str | None = None
    catalog_description: str | None = None
    proof_purpose: str | None = None
    fee_detail: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class StepResponse(BaseModel):
    id: int
    step_name: str
    status: str = "pending"
    progress: int = 0
    duration_ms: int | None = None
    error_message: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None


class EvidenceCaseResponse(BaseModel):
    id: str
    case_name: str
    case_type: str
    is_minor: bool
    status: str
    plaintiff_info: dict[str, Any] = Field(default_factory=dict)
    defendant_info: dict[str, Any] = Field(default_factory=dict)
    catalog_data: dict[str, Any] = Field(default_factory=dict)
    catalog_pdf_path: str | None = None
    analysis_result: dict[str, Any] = Field(default_factory=dict)
    validation_result: dict[str, Any] = Field(default_factory=dict)
    missing_items: dict[str, Any] = Field(default_factory=dict)
    export_bundle_path: str | None = None
    export_files: dict[str, Any] = Field(default_factory=dict)
    lawyer_info: list[dict[str, str]] = Field(default_factory=list, description="律师信息 [{name, phone}]")
    metadata: dict[str, Any] = Field(default_factory=dict)
    materials: list[MaterialResponse] = Field(default_factory=list)
    steps: list[StepResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class EvidenceCaseListResponse(BaseModel):
    items: list[EvidenceCaseResponse]
    total: int


# ─── 材料编辑 ────────────────────────────────────────────────────────────────

class UpdateMaterialRequest(BaseModel):
    manual_category: str | None = Field(default=None, max_length=50)
    catalog_title: str | None = Field(default=None, max_length=200)
    catalog_description: str | None = Field(default=None, max_length=1000)
    proof_purpose: str | None = Field(default=None, max_length=500)
    manual_edit: dict[str, Any] | None = None


class UpdateCaseRequest(BaseModel):
    case_name: str | None = Field(default=None, min_length=1, max_length=500, description="案件名称")
    case_type: Literal["injury", "death"] | None = Field(default=None, description="案件类型")
    is_minor: bool | None = Field(default=None, description="是否未成年人")
    lawyer_info: list[dict[str, str]] | None = Field(default=None, max_length=2, description="律师信息，格式: [{name, phone}, ...]，最多2个")
    defendant_phone: str | None = Field(default=None, max_length=20, description="被告联系电话（手动输入）")

    @field_validator("lawyer_info")
    @classmethod
    def validate_lawyer_info(cls, v: list[dict[str, str]] | None) -> list[dict[str, str]] | None:
        if v is None:
            return v
        for i, lawyer in enumerate(v):
            if "name" not in lawyer or not lawyer["name"].strip():
                raise ValueError(f"律师{i+1}必须包含姓名(name)")
            if "phone" not in lawyer or not lawyer["phone"].strip():
                raise ValueError(f"律师{i+1}必须包含电话(phone)")
            # 验证姓名长度
            if len(lawyer["name"]) > 20:
                raise ValueError(f"律师{i+1}姓名不能超过20个字符")
            # 验证电话格式（中国大陆手机号或座机）
            phone = lawyer["phone"].strip()
            if not re.match(r"^1[3-9]\d{9}$|^0\d{2,3}-?\d{6,8}$", phone):
                raise ValueError(f"律师{i+1}电话格式不正确（应为手机号或座机号）")
        return v

    @field_validator("defendant_phone")
    @classmethod
    def validate_defendant_phone(cls, v: str | None) -> str | None:
        if v is None or v.strip() == "":
            return v
        phone = v.strip()
        # 允许手机号、座机号、或"未知"
        if phone == "未知":
            return phone
        if not re.match(r"^1[3-9]\d{9}$|^0\d{2,3}-?\d{6,8}$", phone):
            raise ValueError("电话格式不正确（应为手机号或座机号，如13800138000或010-12345678）")
        return phone


# ─── 清单编辑 ────────────────────────────────────────────────────────────────

class CatalogItemUpdate(BaseModel):
    material_id: str
    manual_category: str | None = Field(default=None, max_length=50)
    catalog_title: str | None = Field(default=None, max_length=200)
    catalog_description: str | None = Field(default=None, max_length=1000)
    proof_purpose: str | None = Field(default=None, max_length=500)
    sort_order: int | None = None


class UpdateCatalogRequest(BaseModel):
    items: list[CatalogItemUpdate] = Field(default_factory=list)


# ─── 进度响应 ────────────────────────────────────────────────────────────────

class ProgressResponse(BaseModel):
    case_id: str
    status: str
    current_step: str | None = None
    total_steps: int = 0
    completed_steps: int = 0
    progress_percent: float = 0.0
    steps: list[StepResponse] = Field(default_factory=list)


# ─── 清单响应 ────────────────────────────────────────────────────────────────

class CatalogGroupResponse(BaseModel):
    category: str
    category_name: str
    items: list[MaterialResponse]


class CatalogResponse(BaseModel):
    case_id: str
    case_name: str
    case_type: str
    groups: list[CatalogGroupResponse]
    fee_summary: dict[str, Any] = Field(default_factory=dict)
    total_amount: float = 0.0


# ─── 分析响应 ────────────────────────────────────────────────────────────────

class AnalysisResponse(BaseModel):
    case_id: str
    status: str
    analysis_result: dict[str, Any] = Field(default_factory=dict)
    validation_result: dict[str, Any] = Field(default_factory=dict)
    missing_items: dict[str, Any] = Field(default_factory=dict)
    error_message: str | None = Field(default=None, description="分析失败时的错误信息")


# ─── 导出响应 ────────────────────────────────────────────────────────────────

class ExportBundleResponse(BaseModel):
    case_id: str
    message: str
    bundle_path: str | None = None


class ProcessResponse(BaseModel):
    case_id: str
    message: str
    task_id: str | None = None


# ─── 案件列表项（精简） ──────────────────────────────────────────────────────

class EvidenceCaseListItem(BaseModel):
    id: str
    case_name: str
    case_type: str
    is_minor: bool
    status: str
    created_at: datetime
    updated_at: datetime


class EvidenceCaseListSlimResponse(BaseModel):
    items: list[EvidenceCaseListItem]
    total: int


# ─── 赔偿计算相关 Schema ──────────────────────────────────────────────────────

class CompensationParamsUpdate(BaseModel):
    """赔偿计算参数更新请求"""
    annual_income: Optional[Decimal] = None       # 上年度城镇居民人均可支配收入
    annual_consumption: Optional[Decimal] = None   # 上年度城镇居民人均消费支出
    monthly_salary: Optional[Decimal] = None       # 上年度职工月均工资
    daily_food_subsidy: Optional[Decimal] = None   # 住院伙食补助日标准
    daily_nutrition: Optional[Decimal] = None      # 营养费日标准
    compensation_years: Optional[int] = None       # 赔偿年限
    disability_coefficient: Optional[Decimal] = None  # 伤残系数 (0.1-1.0)
    hospital_days: Optional[int] = None            # 住院天数
    lost_wage_days: Optional[int] = None           # 误工天数
    nursing_days: Optional[int] = None             # 护理天数
    nutrition_days: Optional[int] = None           # 营养期天数


class FeeSourceResponse(BaseModel):
    """费用来源素材"""
    material_id: str
    filename: str
    ocr_snippet: Optional[str] = None


class CompensationItemResponse(BaseModel):
    """单项赔偿费用"""
    fee_type: str
    fee_name: str
    amount: Decimal
    manual_amount: Optional[Decimal] = None
    calculation_basis: Optional[str] = None
    is_manual: bool = False
    sources: List[FeeSourceResponse] = []


class CompensationParamsResponse(BaseModel):
    """赔偿计算参数响应"""
    annual_income: Decimal
    annual_consumption: Decimal
    monthly_salary: Decimal
    daily_food_subsidy: Decimal
    daily_nutrition: Decimal
    compensation_years: int
    disability_coefficient: Decimal
    hospital_days: int = 0
    lost_wage_days: int = 0
    nursing_days: int = 0
    nutrition_days: int = 0


class CompensationResponse(BaseModel):
    """赔偿计算结果"""
    case_id: str
    items: List[CompensationItemResponse]
    total_amount: Decimal
    params: CompensationParamsResponse
    calculated_at: Optional[str] = None


class CompensationCalculateRequest(BaseModel):
    """赔偿计算请求"""
    params: Optional[CompensationParamsUpdate] = None


class CompensationItemUpdate(BaseModel):
    """单项赔偿金额手动调整"""
    fee_type: str
    manual_amount: Optional[Decimal] = None


class CompensationUpdateRequest(BaseModel):
    """赔偿数据批量更新请求"""
    items: List[CompensationItemUpdate]
    params: Optional[CompensationParamsUpdate] = None
